File: pylib/qlat/timer.py
import cqlat as c

from cqlat import get_id_node
from cqlat import timer_display, timer_autodisplay
from cqlat import timer_display_stack, timer_display_stack_always
from cqlat import timer_reset

import functools

timer_total_flops = {}

def acc_timer_flops(fname, flops):
    if fname in timer_total_flops:
        timer_total_flops[fname] += flops
    else:
        timer_total_flops[fname] = flops

class Timer:

    def __init__(self, fname, is_verbose = False):
        if fname is None:
            self.cdata = None
        else:
            self.cdata = c.mk_timer(fname)
            self.is_verbose = is_verbose
            acc_timer_flops(fname, 0)

    def __del__(self):
        if self.cdata is None:
            return
        c.free_timer(self)

    def start(self, is_verbose = None):
        if self.cdata is None:
            return
        if is_verbose is None:
            is_verbose = self.is_verbose
        c.start_timer(self, is_verbose)

    def stop(self, is_verbose = None):
        if self.cdata is None:
            return
        if is_verbose is None:
            is_verbose = self.is_verbose
        c.stop_timer(self, is_verbose)

    def set_flops(self, flops):
        if self.cdata is None:
            return
        c.set_flops_timer(self, flops)

def timer(func):
    fname = "py:" + func.__name__
    qtimer = Timer(fname)
    @functools.wraps(func)
    def qtimer_func(*args, **kwargs):
        qtimer.start()
        start_flops = timer_total_flops[fname]
        ret = func(*args, **kwargs)
        stop_flops = timer_total_flops[fname]
        qtimer.set_flops(stop_flops - start_flops)
        qtimer.stop()
        return ret
    return qtimer_func

def timer_verbose(func):
    fname = "py:" + func.__name__
    qtimer = Timer(fname, True)
    @functools.wraps(func)
    def qtimer_func(*args, **kwargs):
        qtimer.start()
        start_flops = timer_total_flops[fname]
        ret = func(*args, **kwargs)
        stop_flops = timer_total_flops[fname]
        qtimer.set_flops(stop_flops - start_flops)
        qtimer.stop()
        return ret
    return qtimer_func

def displayln(*args):
    print(*args, flush=True)

def displayln_info(*args):
    if get_id_node() == 0:
        displayln(*args)

class TimerNone(Timer):

    def __init__(self):
        super().__init__(None)

File: qlat-utils/pylib/cqlat_utils/timer.cpp
#include "lib.h"

#include <qlat-utils/timer.h>

EXPORT(get_id_node, {
  using namespace qlat;
  return py_convert(get_id_node());
})

EXPORT(get_num_node, {
  using namespace qlat;
  return py_convert(get_num_node());
})

EXPORT(mk_timer, {
  using namespace qlat;
  PyObject* p_fname = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_fname)) {
    return NULL;
  }
  std::string fname;
  py_convert(fname, p_fname);
  Timer* ptimer = new Timer(fname, false);
  return py_convert((void*)ptimer);
})

EXPORT(free_timer, {
  using namespace qlat;
  return free_obj<Timer>(args);
})

EXPORT(start_timer, {
  using namespace qlat;
  PyObject* p_timer = NULL;
  bool is_verbose = false;
  if (!PyArg_ParseTuple(args, "O|b", &p_timer, &is_verbose)) {
    return NULL;
  }
  Timer& timer = py_convert_type<Timer>(p_timer);
  timer.start(is_verbose);
  Py_RETURN_NONE;
})

EXPORT(stop_timer, {
  using namespace qlat;
  PyObject* p_timer = NULL;
  bool is_verbose = false;
  if (!PyArg_ParseTuple(args, "O|b", &p_timer, &is_verbose)) {
    return NULL;
  }
  Timer& timer = py_convert_type<Timer>(p_timer);
  timer.stop(is_verbose);
  Py_RETURN_NONE;
})

EXPORT(set_flops_timer, {
  using namespace qlat;
  PyObject* p_timer = NULL;
  long flops = false;
  if (!PyArg_ParseTuple(args, "Ol", &p_timer, &flops)) {
    return NULL;
  }
  Timer& timer = py_convert_type<Timer>(p_timer);
  timer.flops = flops;
  Py_RETURN_NONE;
})

EXPORT(timer_display, {
  using namespace qlat;
  PyObject* p_str = NULL;
  if (!PyArg_ParseTuple(args, "|O", &p_str)) {
    return NULL;
  }
  std::string str;
  if (p_str != NULL) {
    py_convert(str, p_str);
  }
  Timer::display(str);
  Py_RETURN_NONE;
})

EXPORT(timer_autodisplay, {
  using namespace qlat;
  Timer::autodisplay();
  Py_RETURN_NONE;
})

EXPORT(timer_reset, {
  using namespace qlat;
  long max_call_times_for_always_show_info = -1;
  if (!PyArg_ParseTuple(args, "|l", &max_call_times_for_always_show_info)) {
    return NULL;
  }
  Timer::reset(max_call_times_for_always_show_info);
  Py_RETURN_NONE;
})

EXPORT(timer_fork, {
  using namespace qlat;
  long max_call_times_for_always_show_info = -1;
  if (!PyArg_ParseTuple(args, "|l", &max_call_times_for_always_show_info)) {
    return NULL;
  }
  Timer::fork(max_call_times_for_always_show_info);
  Py_RETURN_NONE;
})

EXPORT(timer_merge, {
  using namespace qlat;
  Timer::merge();
  Py_RETURN_NONE;
})

EXPORT(timer_display_stack_always, {
  using namespace qlat;
  Timer::display_stack_always();
  Py_RETURN_NONE;
})

EXPORT(timer_display_stack, {
  using namespace qlat;
  Timer::display_stack();
  Py_RETURN_NONE;
})

EXPORT(verbose_level, {
  using namespace qlat;
  PyObject* p_level = NULL;
  if (!PyArg_ParseTuple(args, "|O", &p_level)) {
    return NULL;
  }
  if (NULL != p_level) {
    const long level = py_convert_data<long>(p_level);
    verbose_level() = level;
  }
  return py_convert(verbose_level());
})

EXPORT(get_actual_start_time, {
  using namespace qlat;
  PyObject* p_time = NULL;
  if (!PyArg_ParseTuple(args, "|O", &p_time)) {
    return NULL;
  }
  if (NULL != p_time) {
    const double time = py_convert_data<double>(p_time);
    get_actual_start_time() = time;
  }
  return py_convert(get_actual_start_time());
})

EXPORT(get_start_time, {
  using namespace qlat;
  PyObject* p_time = NULL;
  if (!PyArg_ParseTuple(args, "|O", &p_time)) {
    return NULL;
  }
  if (NULL != p_time) {
    const double time = py_convert_data<double>(p_time);
    get_start_time() = time;
  }
  return py_convert(get_start_time());
})

EXPORT(get_time, {
  using namespace qlat;
  return py_convert(get_time());
})
